src/shogun/modelselection/NewGridSearch.cpp

```diff
@@ -0,0 +1,891 @@
+	};
+
+	for (auto& param_pair : m_param_mapping)
+	{
+		param = param_pair.first;
+		sg_any_dispatch(
+		    param_pair.second, sg_vector_typemap, shogun::None{}, get_lambda);
+	}
+	return tree;
+}
+
+ParameterNode*
+GridParameters::attach(const std::string& param, ParameterNode* node)
+{
+	auto* gp_node = dynamic_cast<GridParameters*>(node);
+	if (!gp_node)
+		SG_SERROR("Need to pass an object of type GridParameters")
+	auto* param_copy = new GridParameters(*gp_node);
+	if (!ParameterNode::set_param_helper(param, param_copy))
+		SG_SERROR("Could not attach %s", param.c_str());
+	// Updating node list (might) require resetting the node iterators
+	m_current_node = m_nodes.begin();
+	m_current_internal_node = m_current_node->second.begin();
+	return this;
+}
+
+ParameterNode* GridParameters::attach(
+    const std::string& param, const std::shared_ptr<ParameterNode>& node)
+{
+	return attach(param, node.get());
+}
+
+CSGObject* GridSearch::train(CFeatures* features, CLabels* labels, bool verbose)
+{
+	auto parameter_combination = std::unique_ptr<ParameterNode>(get_next());
+	SG_SDEBUG(
+	    "GridSearch::train - "
+	    "CURRENT NODE: %s\n",
+	    parameter_combination->to_string().c_str())
+	auto* machine =
+	    ParameterNode::to_object(parameter_combination)->as<CMachine>();
+	SG_SDEBUG("machine: %d\n", machine->ref_count())
+	m_strategy->put("labels", labels);
+	// TODO: add non lock option
+	auto eval_machine = new CCrossValidation(
+	    machine, features, labels, m_strategy, m_evaluation, false);
+	SG_UNREF(machine)
+
+	std::unique_ptr<ParameterNode> best_combination =
+	    std::move(parameter_combination);
+	auto* best_result = (CCrossValidationResult*)(eval_machine->evaluate());
+
+	// apply all combinations and search for best one
+	while (!is_complete())
+	{
+		// delete old parameter combination and replace with new one
+		SG_SDEBUG("GridSearch::train - GETTING NEW COMBINATION\n")
+		parameter_combination.reset(get_next());
+		SG_SDEBUG("GridSearch::train - GOT NEW COMBINATION\n")
+
+		// the machine is returned with a ref_count of 1 because of clone
+		auto* tmp =
+		    ParameterNode::to_object(parameter_combination)->as<CMachine>();
+
+		// the old machine can now be unref'ed
+		auto* old = eval_machine->get<CMachine*>("machine");
+		SG_SDEBUG(
+		    "old %s: %d\n", eval_machine->get<CMachine*>("machine")->get_name(),
+		    eval_machine->get<CMachine*>("machine")->ref_count())
+		SG_UNREF(old)
+
+		eval_machine->put("machine", tmp);
+		// tmp should only be known by eval_machine
+		SG_UNREF(tmp)
+		SG_SDEBUG("tmp %s: %d\n", tmp->get_name(), tmp->ref_count())
+
+		if (verbose)
+		{
+			SG_SPRINT(
+			    "GridSearch::train - "
+			    "trying combination: %s\n",
+			    parameter_combination->to_string().c_str())
+			SG_SPRINT(
+			    "GridSearch::train - "
+			    "trying model: %s\n",
+			    eval_machine->get<CMachine*>("machine")->to_string().c_str())
+		}
+
+		// note that this may implicitly lock and unlock the machine
```

Comment:
I think we should totally remove this concept of locking machines for now tbh. We could add that back later. It is also not something that would be sitting inside GridSearch, but rather in the cross-validation codes, since that is where the issue of precomputing stuff when re-training multiple times using the same parameters happens